### ISSUE TITLE
`no-useless-spread` should report when passing spreading parameters to `new Set`

### ISSUE THREAD

**Issue by zanminkian** [sindresorhus/eslint-plugin-unicorn#2412]
1️⃣ Explain here what's wrong
```ts
function getNames() {
  return ['foo', 'bar', 'foo'];
}

const uniqueNames = [...new Set(...getNames())]; // I mistakenly spreading the `getNames()`
console.log(uniqueNames) // Print [ 'f', 'o' ]. But what I expect is ['foo', 'bar']
```
`no-useless-spread` should report error on the code `new Set(...getNames())`. But actually not.

Because `new Set()` accept one or zero parameter, we don't need to spread the parameter.

2️⃣ Specify which rule is buggy here and in the title
no-useless-spread

3️⃣ More information
TypeScript should report error when compiling the code above. Unfortunately, this problem havn't been fixed yet. Set https://github.com/microsoft/TypeScript/issues/59390 and https://github.com/microsoft/TypeScript/issues/48575

**Comment by fisker** [comment]
`new Set` take 0~1 arguments https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/Set#parameters

**Comment by fisker** [comment]
> Because new Set() accept one or zero parameter, we don't need to spread the parameters.

This doesn't mean it can't spread the parameters


```js
const foo = [[1,2]]

new Set(...foo)
```

**Comment by zanminkian** [comment]
> This doesn't mean it can't spread the parameters
> 
> ```js
> const foo = [[1,2]]
> 
> new Set(...foo)
> ```

It can, but it's unnecessary.

```ts
const foo = [[1,2]];
new Set(foo[0]);
```

Only the first item will be used in `new Set(...foo)`. So Just use `new Set(foo[0])`.

Pass spreading parameters to `new Set` is a bad code style.
- It's confused.
- Not type safe enough (due to TypeScript bug).

**Comment by fregante** [comment]
There's a slight difference I think. `no-useless-spread` suggests that you "don't need it", while here it leads to a different result altogether, probably not what you wanted in the first place. 

However I still think it makes sense to use it here. The spread only works correctly when `foo.length <= 1`.

```js
const foo = []; // Empty
new Set(...foo) == new Set() == new Set(undefined); // Ok, works just as well
```

```js
const foo = [[1]]; // One
new Set(...foo) == new Set(foo[0]); // Ok, works just as well
```

```js
const foo = [[1], [2]]; // Two
new Set(...foo) == new Set(foo[0], foo[1]); // ❌ Item 2 lost
```

**Comment by zanminkian** [comment]
> ```js
> const foo = [[1], [2]]; // Two
> new Set(...foo) == new Set(foo[0], foo[1]); // ❌ Item 2 lost
> ```
> 
> The only caution would be to _not_ autofix it in this case.

```js
const foo = [[1], [2]]; // Two
new Set(...foo) == new Set(foo[0], foo[1]) == new Set(foo[0]); // Ok, works just as well
```

Auto fixing is safe. Just confidently change all `new Set(...foo)` to `new Set(foo[0])`.

**Comment by fregante** [comment]
That only works if you know the type of `foo`. If it’s an iterator other than Array that breaks it.

**Comment by zanminkian** [comment]
My fault.

```js
const map = new Map([['a','b']]);
new Set(...map) != new Set(map[0]);
```
